File: frontend/types/order.ts
// Order Types

export enum OrderType {
  DINE_IN = 'DINE_IN',
  TAKEAWAY = 'TAKEAWAY',
  DELIVERY = 'DELIVERY',
}

export enum OrderStatus {
  PENDING = 'PENDING',
  IN_PREPARATION = 'IN_PREPARATION',
  READY_FOR_DELIVERY = 'READY_FOR_DELIVERY',
  IN_DELIVERY = 'IN_DELIVERY',
  COMPLETED = 'COMPLETED',
  CANCELLED = 'CANCELLED',
}

export enum PaymentMethod {
  CASH = 'CASH',
  CREDIT_CARD = 'CREDIT_CARD',
  DEBIT_CARD = 'DEBIT_CARD',
  PIX = 'PIX',
  MERCADO_PAGO = 'MERCADO_PAGO',
  OTHER = 'OTHER',
}

export enum PaymentStatus {
  PENDING = 'PENDING',
  PAID = 'PAID',
  FAILED = 'FAILED',
  REFUNDED = 'REFUNDED',
  PARTIALLY_REFUNDED = 'PARTIALLY_REFUNDED',
}

export enum OrderPriority {
  LOW = 'LOW',
  NORMAL = 'NORMAL',
  HIGH = 'HIGH',
  URGENT = 'URGENT',
}

// Order interfaces
export interface Order {
  id: string;
  orderNumber: string;
  tenantId: string;
  customerId?: string;
  customerName?: string;
  customerPhone?: string;
  userId?: string;
  userName?: string;
  orderType: OrderType;
  status: OrderStatus;
  tableNumber?: string;
  items: OrderItem[];
  subtotal: number;
  discount: number;
  deliveryFee: number;
  taxAmount: number;
  total: number;
  paymentMethod?: PaymentMethod;
  paymentStatus: PaymentStatus;
  paymentReference?: string;
  isTakeaway: boolean;
  priority: OrderPriority;
  estimatedPrepTimeMinutes: number;
  notes?: string;
  idempotencyKey?: string;
  createdAt: string;
  updatedAt: string;
  completedAt?: string;
  cancelledAt?: string;
  cancelledReason?: string;
}

export interface OrderItem {
  id: string;
  orderId: string;
  productId: string;
  productSku: string;
  productName: string;
  quantity: number;
  unitPrice: number;
  totalPrice: number;
  notes?: string;
  status: OrderItemStatus;
  preparationStartedAt?: string;
  preparationCompletedAt?: string;
  displayOrder: number;
  isCombo: boolean;
  parentItemId?: string;
  customizations?: ProductCustomization[];
}

export enum OrderItemStatus {
  PENDING = 'PENDING',
  IN_PREPARATION = 'IN_PREPARATION',
  READY = 'READY',
  COMPLETED = 'COMPLETED',
  CANCELLED = 'CANCELLED',
}

export interface ProductCustomization {
  ingredientId: string;
  ingredientName: string;
  action: CustomizationAction;
  quantity?: number;
  extraPrice?: number;
}

export enum CustomizationAction {
  ADD = 'ADD',
  REMOVE = 'REMOVE',
  REPLACE = 'REPLACE',
}

// Order Request types
export interface CreateOrderRequest {
  tenantId: string;
  customerId?: string;
  orderType: OrderType;
  tableNumber?: string;
  items: CreateOrderItemRequest[];
  notes?: string;
  paymentMethod?: PaymentMethod;
  isTakeaway: boolean;
  priority?: OrderPriority;
  idempotencyKey?: string;
}

export interface CreateOrderItemRequest {
  productId: string;
  quantity: number;
  notes?: string;
  customizations?: CreateProductCustomizationRequest[];
}

export interface CreateProductCustomizationRequest {
  ingredientId: string;
  action: CustomizationAction;
  quantity?: number;
}

// Order Response types
export interface OrderResponse extends Order {}

export interface OrderSummary {
  id: string;
  orderNumber: string;
  tenantId: string;
  customerName?: string;
  orderType: OrderType;
  status: OrderStatus;
  total: number;
  paymentStatus: PaymentStatus;
  createdAt: string;
  tableNumber?: string;
}

export interface OrderListResponse {
  content: OrderSummary[];
  page: number;
  size: number;
  totalElements: number;
  totalPages: number;
  last: boolean;
}

// Order filters
export interface OrderFilters {
  tenantId: string;
  status?: OrderStatus[];
  orderType?: OrderType[];
  paymentStatus?: PaymentStatus[];
  startDate?: string;
  endDate?: string;
  customerId?: string;
  page?: number;
  size?: number;
  sortBy?: string;
  sortDirection?: 'ASC' | 'DESC';
}

// Order metrics
export interface OrderMetrics {
  totalOrders: number;
  totalRevenue: number;
  averageOrderValue: number;
  completedOrders: number;
  pendingOrders: number;
  cancelledOrders: number;
  ordersByStatus: Record<OrderStatus, number>;
  ordersByType: Record<OrderType, number>;
  revenueByDay: Record<string, number>;
}

File: backend/src/main/kotlin/com/burgerflow/model/Order.kt
package com.burgerflow.model

import jakarta.persistence.*
import java.math.BigDecimal
import java.time.LocalDateTime
import java.util.UUID

@Entity
@Table(name = "orders")
data class Order(
    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    val id: UUID? = null,
    
    @Column(nullable = false)
    var tenantId: UUID,
    
    @Column(name = "order_number", nullable = false, unique = true)
    var orderNumber: String,
    
    @Column(nullable = false)
    var customerId: UUID? = null,
    
    @Column(nullable = false)
    var userId: UUID? = null,
    
    @Column(nullable = false)
    @Enumerated(EnumType.STRING)
    var orderType: OrderType = OrderType.DINE_IN,
    
    @Column(nullable = false)
    @Enumerated(EnumType.STRING)
    var status: OrderStatus = OrderStatus.PENDING,
    
    @Column(name = "table_number")
    var tableNumber: String? = null,
    
    @Column(name = "notes")
    var notes: String? = null,
    
    @Column(name = "subtotal", nullable = false, precision = 12, scale = 2)
    var subtotal: BigDecimal = BigDecimal.ZERO,
    
    @Column(name = "discount", precision = 12, scale = 2)
    var discount: BigDecimal = BigDecimal.ZERO,
    
    @Column(name = "delivery_fee", precision = 12, scale = 2)
    var deliveryFee: BigDecimal = BigDecimal.ZERO,
    
    @Column(name = "tax_amount", precision = 12, scale = 2)
    var taxAmount: BigDecimal = BigDecimal.ZERO,
    
    @Column(name = "total", nullable = false, precision = 12, scale = 2)
    var total: BigDecimal = BigDecimal.ZERO,
    
    @Column(name = "payment_method")
    @Enumerated(EnumType.STRING)
    var paymentMethod: PaymentMethod? = null,
    
    @Column(name = "payment_status")
    @Enumerated(EnumType.STRING)
    var paymentStatus: PaymentStatus = PaymentStatus.PENDING,
    
    @Column(name = "payment_reference")
    var paymentReference: String? = null,
    
    @Column(name = "is_takeaway", nullable = false)
    var isTakeaway: Boolean = false,
    
    @Column(name = "priority", nullable = false)
    @Enumerated(EnumType.STRING)
    var priority: OrderPriority = OrderPriority.NORMAL,
    
    @Column(name = "estimated_prep_time_minutes", nullable = false)
    var estimatedPrepTimeMinutes: Int = 15,
    
    @Column(name = "idempotency_key", unique = true)
    var idempotencyKey: String? = null,
    
    @Column(name = "created_at", nullable = false, updatable = false)
    val createdAt: LocalDateTime = LocalDateTime.now(),
    
    @Column(name = "updated_at")
    var updatedAt: LocalDateTime = LocalDateTime.now(),
    
    @Column(name = "completed_at")
    var completedAt: LocalDateTime? = null,
    
    @Column(name = "cancelled_at")
    var cancelledAt: LocalDateTime? = null,
    
    @Column(name = "cancelled_reason")
    var cancelledReason: String? = null
) {
    @PreUpdate
    fun preUpdate() {
        updatedAt = LocalDateTime.now()
    }
    
    fun calculateTotal(): BigDecimal {
        return subtotal - discount + deliveryFee + taxAmount
    }
    
    fun isPaid(): Boolean {
        return paymentStatus == PaymentStatus.PAID
    }
    
    fun isCompleted(): Boolean {
        return status == OrderStatus.COMPLETED
    }
    
    fun isCancelled(): Boolean {
        return status == OrderStatus.CANCELLED
    }
    
    fun canBeCancelled(): Boolean {
        return status in listOf(OrderStatus.PENDING, OrderStatus.IN_PREPARATION)
    }
}

enum class OrderType {
    DINE_IN,
    TAKEAWAY,
    DELIVERY
}

enum class OrderStatus {
    PENDING,
    IN_PREPARATION,
    READY_FOR_DELIVERY,
    IN_DELIVERY,
    COMPLETED,
    CANCELLED
}

enum class PaymentMethod {
    CASH,
    CREDIT_CARD,
    DEBIT_CARD,
    PIX,
    MERCADO_PAGO,
    OTHER
}

enum class PaymentStatus {
    PENDING,
    PAID,
    FAILED,
    REFUNDED,
    PARTIALLY_REFUNDED
}

enum class OrderPriority {
    LOW,
    NORMAL,
    HIGH,
    URGENT
}
